test/lib/independence/independent_request.test.js

```diff
@@ -0,0 +1,146 @@
+const chai = require('chai')
+const sinon = require('sinon')
+const spies = require('chai-spies')
+const proxyquire = require('proxyquire')
+const Promise = require('bluebird')
+
+chai.use(spies)
+
+const mockStorage = {
+  put: (key, value) => { mockStorage._data[key] = value },
+  get: key => mockStorage._data[key],
+  delete: key => { delete mockStorage._data[key] },
+  all: () => mockStorage._data,
+  _data: {}
+}
+const getmockStorage = chai.spy(() => mockStorage)
+const MOCK_DATE = 1558625994627
+
+const { IndependentRequest } = proxyquire('../../../lib/independence/independent_request', {
+  './local_storage_namespace': {
+    LocalStorageNamespace: getmockStorage
+  }
+})
+
+const expect = chai.expect
+
+describe('independent request', () => {
+  let clock
+  const sandbox = chai.spy.sandbox()
+
+  beforeEach(() => {
+    mockStorage._data = { }
+    sandbox.on(mockStorage, ['put', 'get', 'all'])
+    clock = sinon.useFakeTimers(MOCK_DATE)
+  })
+
+  afterEach(() => {
+    sandbox.restore()
+    clock.restore()
+  })
+
+  it('created a storage with the key `independentRequests`', () => {
+    expect(getmockStorage).to.have.been.called.with('independentRequests')
+    expect(IndependentRequest.storage).to.equal(mockStorage)
+  })
+
+  it('defaultly sets the key to be the current date', () => {
+    expect(new IndependentRequest().key).to.equal(MOCK_DATE)
+  })
+
+  it('calls the adapter with the given data', () => {
+    const mockAdapter = chai.spy(() => Promise.resolve({ status: 200 }))
+    const data = { }
+    const request = new IndependentRequest(data)
+
+    return request.send(mockAdapter).then(() => {
+      expect(mockAdapter).to.have.been.called.with(data)
+    })
+  })
+
+  it('saves the request if it fails with status 500', () => {
+    const error = new Error('some error')
+    error.status = 500
+    const mockAdapter = () => Promise.resolve().then(() => { throw error })
+    const key = 'key'
+    const data = { }
+    const request = new IndependentRequest(data, key)
+
+    return request.send(mockAdapter)
+      .then(() => { expect(false).to.equal(true) })
```

Comment:
1. remember that assertions uses exception. meaning this line is throwing exception that is going to be swallowed by  the next `catch` in the promise chain. use the less common form of `then`  which is `.then(regularCallback, ErrorCallback)` or use `chai-as-promised` and make an assertion that a promise will be rejected.
3. isn't this test check 2 things: a. the request fails, b. the data have been saved.
2. use `except.fail()` instead of the current assertation

Reply:
About No. 2: the tests checks that `if` the request fails, the data has been saved. The failing of the request should be a part of the `act`. I'm asserting it here, so that someone doesn't accedentily change it so the request would succeed, and then the test would still work.
I'll fix the other two

Reply:
I see it as two tests because the two tests are:
1. independent request reject sending process when adapter fails
2. independent request saves the request when adapter fails

Therefore if someone changes the `send` method such that it won't fail when adapter fails the test would immediately say what is the problem.

Also for the second test, put the assertion after an empty catch method and not inside it. This way the assertion will be checked regardless of if the `send` method rejected or not. This is good because if we decide that we don't want the send method to fail when the adapter fails we only changes one test.

**Proper unit tests should fail for exactly one reason**